src/goto-instrument/contracts/dynamic-frames/dfcc.cpp

```diff
@@ -0,0 +1,497 @@
+
+    PRECONDITION_WITH_DIAGNOSTICS(
+      dsl_contract_handler.pure_contract_symbol_exists(pair.second),
+      "Contract to check '" + id2string(pair.second) + "' not found");
+
+    PRECONDITION_WITH_DIAGNOSTICS(
+      pair.first != harness_id,
+      "Function '" + id2string(pair.first) +
+        "' cannot be both be checked against a contract and be the harness");
+
+    PRECONDITION_WITH_DIAGNOSTICS(
+      pair.second != harness_id,
+      "Function '" + id2string(pair.first) +
+        "' cannot be both the contract to check and be the harness");
+
+    PRECONDITION_WITH_DIAGNOSTICS(
+      to_replace.find(pair.first) == to_replace.end(),
+      "Function '" + id2string(pair.first) +
+        "' cannot be both checked against contract and replaced by a contract");
+
+    PRECONDITION_WITH_DIAGNOSTICS(
+      !instrument.do_not_instrument(pair.first),
+      "CPROVER function or builtin '" + id2string(pair.first) +
+        "' cannot be checked against a contract");
+  }
+
+  for(const auto &pair : to_replace)
+  {
+    // for functions to replace with contracts we don't require the replaced
+    // function to have a body because only the contract is actually used
+    PRECONDITION_WITH_DIAGNOSTICS(
+      utils.function_symbol_exists(pair.first),
+      "Function to replace '" + id2string(pair.first) + "' not found");
+
+    PRECONDITION_WITH_DIAGNOSTICS(
+      dsl_contract_handler.pure_contract_symbol_exists(pair.second),
+      "Contract to replace '" + id2string(pair.second) + "' not found");
+
+    PRECONDITION_WITH_DIAGNOSTICS(
+      pair.first != harness_id,
+      "Function '" + id2string(pair.first) +
+        "' cannot both be replaced with a contract and be the harness");
+
+    PRECONDITION_WITH_DIAGNOSTICS(
+      pair.second != harness_id,
+      "Function '" + id2string(pair.first) +
+        "' cannot both be the contract to use for replacement and be the "
+        "harness");
+  }
+}
+
+void dfcct::partition_function_symbols(
+  std::set<irep_idt> &contract_symbols,
+  std::set<irep_idt> &other_symbols)
+{
+  auto &symbol_table = goto_model.symbol_table;
+
+  // collect contract and other sybols
+  for(auto &entry : symbol_table)
+  {
+    const symbolt &symbol = entry.second;
+
+    // not a function symbol
+    if(symbol.type.id() != ID_code)
+      continue;
+
+    // is it a pure contract ?
+    const irep_idt &sym_name = symbol.name;
+    if(symbol.is_property && has_prefix(id2string(sym_name), "contract::"))
+    {
+      contract_symbols.insert(sym_name);
+    }
+    else
+    {
+      // it is not a contract
+      other_symbols.insert(sym_name);
+    }
+  }
+}
+
+void dfcct::transform_goto_model(
+  const irep_idt &harness_id,
+  const std::map<irep_idt, irep_idt> &to_check,
+  const bool allow_recursive_calls,
+  const std::map<irep_idt, irep_idt> &to_replace,
+  const bool apply_loop_contracts,
+  const std::set<std::string> &to_exclude_from_nondet_static)
+{
```

Comment:
The body of this function has more than 250 lines. It would really help readability if it could be broken up in a very small number of calls to functions the names of which clearly describe what is going on. Your various `log.status()` outputs provide a good indicator for where splits might be done.

Reply:
split in aux methods